### PR TITLE
Fix log writing failed. can't be called from trap context (thread version)

### PR DESCRIPTION
**Pull request by sonots**
Fix https://github.com/resque/resque/issues/1493

### CODE REVIEW

**Comment by sonots** [comment]
CI failed, but it is not related with this commit (failed with ruby 1.9.3 and 1.8.7)

**Comment by fw42** [comment]
Hmmm.. I'm not a fan of how this will make your application multi-threaded in possibly unexpected ways. For example, at Shopify, our Logger is actually logging to Kafka, and with this PR, that Kafka logger would run in a seperate thread. Now I have to think about if the logging code is thread-safe, which I wouldn't expect having to think about in an otherwise single-threaded application.

@dylanahsmith, how do you feel about this?

File: lib/resque/signal_handler.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  def trap(sig, &block)
+    # normalize signal names
+    sig = sig.to_s.upcase
+    if sig[0,3] == "SIG"
+      sig = sig[3..-1]
+    end
+
+    @handlers[sig] = block
+    Kernel.trap(sig) { @self_write.puts(sig) }
```
Comment (dylanahsmith) [inline lib/resque/signal_handler.rb]: Isn't it the logging not the signal handling that needs to be done in another thread?

**Comment by dylanahsmith** [comment]
If you want a separate thread to do the logging, then couldn't that be accomplished with a custom logger that writes the log entries to a pipe and has its own separate thread for doing the actual logging?

**Comment by sonots** [comment]
Basically, mutex#lock can't be called from trap context. ref. https://bugs.ruby-lang.org/issues/7917
Problem is not only for logger.
Pushing signal into an array or pipe in trap context, and hanling them in somewhere else is the common way handling signal as unicorn, sidekiq, fluentd does.

**Comment by sonots** [comment]
* sidekiq https://github.com/mperham/sidekiq/blob/b6de0bae6b86d5abae25852141768d8ecc8ddedf/lib/sidekiq/cli.rb#L53
* unicorn https://github.com/defunkt/unicorn/blob/master/lib/unicorn/http_server.rb#L120

**Comment by sonots** [comment]
We basically have to avoid heavy work in trap context because we can not trap a new signal during processing something inside trap context.
We therefore should only put signals into an array or pipe in trap context, and then treat signals somewhere else sequentially not to lose new signals.

**Comment by sonots** [comment]
I came up with a way not to use thread > https://github.com/resque/resque/compare/master...sonots:fix_signal_trap2?expand=1

However, in this way, the behavior of signal handler for a forked child must have been changed if TERM_CHILD environment variable is not given.

Before:

* TERM: Set `@shutdown = true`, but actually affects nothing
* INT: Set `@shutdown = true`, but actually affects nothing
* QUIT: Set `@shutdown = true`, but actually affects nothing
* USR1: Do nothing
* USR2: run_hook :before_pause and set `@paused = true`
* CONT: Set `@paused = false` and run_hook :after_pause

After:

* All signals are reverted to DEFAULT

Which do you prefer?

BTW: Rather, I did not get why resque was not calling unregister_signal_handlers for a forked child always. I felt as `Before` behavior was wierd because it disabled TERM or INT signals.

**Comment by fw42** [comment]
The problem I see with your other approach is that that code (where the `sleep` was) will only get executed if the queue is empty (`work_one_job` returns `nil`/`false`). So if there is sufficiently many jobs being pushed into the queue, you will never get to the signal handling part.

Can you open a PR for that so that we can discuss the code?

**Comment by sonots** [comment]
Okay, sent another PR https://github.com/resque/resque/pull/1502

**Comment by sonots** [comment]
I added Thread.stop to stop main thread on receiving signal once, not to require handling a signal to be thread-safe anymore > https://github.com/resque/resque/pull/1499/commits/8fff907d49b73602211ef5efbfdd013053848889

File: lib/resque/worker.rb [hunk 2]
```diff
@@ -861,7 +862,11 @@ def perform_with_fork(job, &block)
 
       begin
         @child = fork do
-          unregister_signal_handlers if term_child
+          if term_child
+            unregister_signal_handlers
+          else
+            @signal_handler.reopen
```
Comment (fw42) [inline lib/resque/worker.rb]: why is this necessary?

Comment (sonots) [inline lib/resque/worker.rb]: It is because

1. fork does not copy thread
2. IO.pipe is shared with the parent process

which results in that receiving a signal in child process makes an effect to its parent process.
`reopen` newly creates signal handler thread and pipe for the child process.

Comment (fw42) [inline lib/resque/worker.rb]: Ok makes sense!

Comment (fw42) [inline lib/resque/worker.rb]: are any tests failing if you remove this line here? If not, can you add a new test?

Comment (sonots) [inline lib/resque/worker.rb]: It seems difficult to create tests for handling signals, but let me try it later.

Comment (sonots) [inline lib/resque/worker.rb]: Added https://github.com/resque/resque/pull/1499/commits/885d51fb86651be61bdbb4485bbe7b27df405331, but sleep 0.1 * 10 times would be feasible...

File: test/child_killing_test.rb [hunk 3]
```diff
@@ -92,11 +98,26 @@ def assert_child_not_running(child_pid)
...
+
+    Process.kill('TERM', child_pid)
+    10.times do
+      assert_nil Process.waitpid(worker_pid, Process::WNOHANG) # should not die
+      sleep 0.1
+    end
+
+    Process.kill('TERM', worker_pid)
+    Process.waitpid(worker_pid)
+    result = Resque.redis.lpop('sigterm-test:result')
```
Comment (fw42) [inline test/child_killing_test.rb]: what does this line do? isn't there an assertion missing here?

File: test/child_killing_test.rb [hunk 4]
```diff
@@ -92,11 +98,26 @@ def assert_child_not_running(child_pid)
...
-      worker_pid, child_pid, result = start_worker(0, true)
+      worker_pid, child_pid, result = start_worker_and_kill_term(0, true)
       assert_equal worker_pid, child_pid, "child_pid should equal worker_pid, since we are not forking"
       assert Resque.redis.lpop( 'sigterm-test:ensure_block_executed' ), 'post cleanup did not occur. SIGKILL sent too early?'
     ensure
       ENV['FORK_PER_JOB'] = old_job_per_fork
     end
   end
+
+  it 'sending signal to a forked child should not make a effect to its parent' do
```
Comment (fw42) [inline test/child_killing_test.rb]: just to double-check: This test fails without the line above?

Comment (sonots) [inline test/child_killing_test.rb]: yes

File: test/child_killing_test.rb (hunk 3 above)
Comment (sonots) [inline test/child_killing_test.rb]: it seems that LongRunningJob does `rpush`, this line is just for `lpop` it. Is it not needed?

**Comment by fw42** [comment]
I like this approach. @dylanahsmith, @steveklabnik, wdyt?

**Comment by sonots** [comment]
> I added Thread.stop to stop main thread on receiving signal once, not to require handling a signal to be thread-safe anymore > 8fff907

This way was not good. Main thread would already be a stop state before calling 
Thread.stop in some reasons. Main thread stucks if Thread.main.wakeup is called before
Thread.stop. (This situation happened to me). Reverted.

**Comment by dylanahsmith** [comment]
I still dislike this approach.  There seems to be an assumption that we need to move all the signal handling logic outside of the signal handler, but 

> Basically, mutex#lock can't be called from trap context. ref. https://bugs.ruby-lang.org/issues/7917
> Problem is not only for logger.

Where else do we use mutex#lock other than in the implementation of the logger?  The signal handlers are not actually that complicated.  Some of those mostly just set an instance variable, which seems like overkill to do by piping to another thread.

The other possible concern might be the before_pause and after_pause hooks, but moving that to a separate thread could break applications using that hook.  If the application wants to avoid this problem, then they could also avoid logging or just delay logging.

What if we just append those messages to an array, which is safe to do in a signal handler, then pop them off and write them in the main thread.  For instance we could do that when methods like Resque::Worker#shutdown? or Resque::Worker#paused? are called to check those instance variables so that we log before the signal actually has an effect.

**Comment by chrisccerami** [comment]
Is this PR still in progress?

**Comment by dchersey** [comment]
Sorry, did this PR get merged?
